### ISSUE TITLE
any suggestion in evaluation mode?

### ISSUE THREAD

**Issue by sebastianpinedaar** [learnables/learn2learn#181]
I am meta-learning a Fully Convolutional Net with some drop-out layers and batch-norm layers. It is known that the training and inference of these layers change. Normally, in Pytorch, we set the model.eval() to configure the inference procedure. However, how does it work in meta-learning? Should we set the evaluation mode after meta-learning and fine-tuning the parameters?

Thanks a lot!

**Comment by seba-1511** [comment]
I think calling `model.eval()` before you adapt to your validation / testing tasks should work. If you try both, could you post your results here?

**Comment by sebastianpinedaar** [comment]
I tried the following two options:

```Python

maml = l2l.algorithms.MAML(model, lr=0.1)
opt = torch.optim.SGD(maml.parameters(), lr=0.001)
for iteration in range(10):
    opt.zero_grad()
    task_model = maml.clone()  # torch.clone() for nn.Modules
    task_model.train()
    adaptation_loss = compute_loss(task_model, train_tasks)
    task_model.adapt(adaptation_loss)  # computes gradient, update task_model in-place
    evaluation_loss = compute_loss(task_model, train_tasks)
    evaluation_loss.backward()  # gradients w.r.t. maml.parameters()
    opt.step()

    task_model = maml.clone()  # torch.clone() for nn.Modules
    task_model.eval()
    adaptation_loss = compute_loss(task_model, test_tasks)
    task_model.adapt(adaptation_loss)  # computes gradient, update task_model in-place
    evaluation_loss = compute_loss(task_model, test_tasks)

```
and:

```Python

maml = l2l.algorithms.MAML(model, lr=0.1)
opt = torch.optim.SGD(maml.parameters(), lr=0.001)
for iteration in range(10):
    opt.zero_grad()
    task_model = maml.clone()  # torch.clone() for nn.Modules
    task_model.train()
    adaptation_loss = compute_loss(task_model, train_tasks)
    task_model.adapt(adaptation_loss)  # computes gradient, update task_model in-place
    evaluation_loss = compute_loss(task_model)
    evaluation_loss.backward()  # gradients w.r.t. maml.parameters()
    opt.step()

    task_model = maml.clone().module  # torch.clone() for nn.Modules
    task_model.train()
    opt2 = torch.optim.SGD(task_model.parameters(), lr=0.01)
    for step in range(adaptation_steps):
       opt2.zero_grad()
       fine_tuning_loss = loss(task_model,  test_tasks)
       fine_tuning_loss.backward()
       opt2.step()
    model.eval()
    evaluation_loss = compute_loss(task_model, test_tasks)

```


Both options gave worse results compared to doing nothing (model.train in any case). The model is a FCN with batch norm layers. Moreover, my data is time series. I suspect that in time series data (or streaming data), the target tasks may change so much that the batch norm decreases the performance in the way that traditional inference is done with this layer. There is an interesting paper studying normalization in tasks for emta-learning: https://arxiv.org/abs/2003.03284 (TaskNorm: Rethinking Batch Normalization for Meta-Learning).

**Comment by seba-1511** [comment]
Thanks! Closing this since it seems we find a solution.
